File: educational-core/src/com/jetbrains/edu/learning/checkio/checker/CheckiOTaskChecker.java
package com.jetbrains.edu.learning.checkio.checker;

import com.intellij.openapi.application.ex.ApplicationUtil;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.project.Project;
import com.jetbrains.edu.learning.checker.CheckResult;
import com.jetbrains.edu.learning.checker.EnvironmentChecker;
import com.jetbrains.edu.learning.checker.TaskChecker;
import com.jetbrains.edu.learning.checker.details.CheckDetailsView;
import com.jetbrains.edu.learning.checkio.connectors.CheckiOOAuthConnector;
import com.jetbrains.edu.learning.courseFormat.CheckStatus;
import com.jetbrains.edu.learning.courseFormat.tasks.EduTask;
import org.jetbrains.annotations.NotNull;

public class CheckiOTaskChecker extends TaskChecker<EduTask> {
  private final CheckiOMissionCheck myMissionCheck;
  private final EnvironmentChecker myEnvChecker;

  public CheckiOTaskChecker(
    @NotNull EduTask task,
    @NotNull EnvironmentChecker envChecker,
    @NotNull Project project,
    @NotNull CheckiOOAuthConnector oAuthConnector,
    @NotNull String interpreterName,
    @NotNull String testFormTargetUrl
  ) {
    super(task, project);

    myMissionCheck = new CheckiOMissionCheck(
      task,
      project,
      oAuthConnector,
      interpreterName,
      testFormTargetUrl
    );
    myEnvChecker = envChecker;
  }

  @NotNull
  @Override
  public CheckResult check(@NotNull ProgressIndicator indicator) {
    try {
      String possibleError = myEnvChecker.checkEnvironment(project);
      if (possibleError != null) return new CheckResult(CheckStatus.Unchecked, possibleError);

      final CheckResult checkResult =
        ApplicationUtil.runWithCheckCanceled(myMissionCheck, ProgressManager.getInstance().getProgressIndicator());

      if (checkResult.getStatus() != CheckStatus.Unchecked) {
        CheckDetailsView.getInstance(project).showJavaFXResult("CheckiO Response", myMissionCheck.getBrowserPanel());
      }

      return checkResult;
    }
    catch (Exception e) {
      LOG.warn(e.getMessage());
      return CheckResult.getFAILED_TO_CHECK();
    }
  }
}

File: educational-core/src/com/jetbrains/edu/learning/checkio/checker/CheckiOTaskChecker.kt
package com.jetbrains.edu.learning.checkio.checker

import com.intellij.openapi.application.ex.ApplicationUtil
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.project.Project
import com.jetbrains.edu.learning.EduSettings
import com.jetbrains.edu.learning.JavaUILibrary
import com.jetbrains.edu.learning.checker.CheckResult
import com.jetbrains.edu.learning.checker.CheckResult.Companion.failedToCheck
import com.jetbrains.edu.learning.checker.EnvironmentChecker
import com.jetbrains.edu.learning.checker.TaskChecker
import com.jetbrains.edu.learning.checker.details.CheckDetailsView.Companion.getInstance
import com.jetbrains.edu.learning.checkio.connectors.CheckiOOAuthConnector
import com.jetbrains.edu.learning.courseFormat.CheckStatus
import com.jetbrains.edu.learning.courseFormat.tasks.EduTask
import com.jetbrains.edu.learning.messages.EduCoreBundle
import org.jetbrains.annotations.NonNls

class CheckiOTaskChecker(
  task: EduTask,
  private val envChecker: EnvironmentChecker,
  project: Project,
  oAuthConnector: CheckiOOAuthConnector,
  @NonNls interpreterName: String,
  @NonNls testFormTargetUrl: String
) : TaskChecker<EduTask>(task, project) {

  private val missionCheck: CheckiOMissionCheck = when (EduSettings.getInstance().javaUiLibraryWithCheck) {
    JavaUILibrary.JCEF -> {
      getJCEFCheckiOMissionCheck(project, task, oAuthConnector, interpreterName, testFormTargetUrl)
      ?: throw RuntimeException("Can't get JCEF CheckiO checker")
    }
    else -> JavaFxCheckiOMissionCheck(project, task, oAuthConnector, interpreterName, testFormTargetUrl)
  }

  override fun check(indicator: ProgressIndicator): CheckResult {
    return try {
      val possibleError = envChecker.getEnvironmentError(project, task)
      if (possibleError != null) {
        return possibleError
      }

      val checkResult = ApplicationUtil.runWithCheckCanceled(missionCheck, ProgressManager.getInstance().progressIndicator)

      if (checkResult.status != CheckStatus.Unchecked) {
        getInstance(project).showResult(EduCoreBundle.message("tab.title.checkio.response"), missionCheck.getPanel())
      }
      checkResult
    }
    catch (e: Exception) {
      LOG.warn(e.message)
      failedToCheck
    }
  }
}